### PR TITLE
Create an example directory and add some code examples.

### PR DESCRIPTION
**Pull request by dota17**
Personally speaking, I think the original code examples for this project are not very easy to find, and the readability is not very high,so I create a example directory and add some code examples here.

### CODE REVIEW

File: example/stringWrite/stringWrite.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "action" : "run",
+    "data" :
+    {
+        "number" : 11
+    }
+}
+
+*/
+
+using namespace std ;
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: Run clang format, and issues here like extra space before the semicolons will disappear.

Also, I would generally prefer no "using namespace std" declarations, even for example code.

File: example/stringWrite/stringWrite.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    string json_file = writer.write(root);
+#else  // new way
+	Json::StreamWriterBuilder builder;
+    string json_file = Json::writeString(builder,root);
+#endif
+
+    cout<< json_file << endl;
+
+    return 0;
+
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: Nit: remove extra line.

File: example/streamWrite/streamWrite.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+>g++ streamWrite.cpp -ljsoncpp -std=c++11 -o streamWrite
+>./streamWrite
+
+{
+    "Age" : 20,
+    "Name" : "robin"
+}
+
+*/
+using namesp std ;
```
Comment (baylesj) [inline example/streamWrite/streamWrite.cpp]: Remove erroneous line. Does this even compile?

File: example/streamWrite/streamWrite.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#incldue <iostream>
```
Comment (baylesj) [inline example/streamWrite/streamWrite.cpp]: Fix build issues with example code. This shouldn't compile.

File: example/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+   )
+
+add_definitions(-D_GLIBCXX_USE_CXX11_ABI)
+set_property(DIRECTORY PROPERTY COMPILE_OPTIONS ${EXTRA_CXX_FLAGS})
+
+IF("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -fpermissive -g -fjsoncpp")
+else()	
+	add_definitions(-D_SCL_SECURE_NO_WARNINGS
+		D_CRT_SECURE_NO_WARNINGS
```
Comment (baylesj) [inline example/CMakeLists.txt]: Nit: fix alignment. The tabbing is all weird here.

File: example/readFromStream/readFromStream.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+{
+    // comment before
+    "key" : "value"
+}
+// comment after
+// comment tail
+*/
+int main(int argc, char* argv[]) {
+  Json::Value jsonRoot;
+  Json::Value jsonItem;
```
Comment (BillyDonahue) [inline example/readFromStream/readFromStream.cpp]: unused?

File: example/readFromStream/readFromStream.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  std::ifstream ifs;
+  ifs.open(argv[1]);
+  jsonRoot.clear();
+
+  Json::CharReaderBuilder builder;
+  builder["collectComments"] = true;
+  JSONCPP_STRING errs;
+  if (!parseFromStream(builder, ifs, &jsonRoot, &errs)) {
+    // std::cout << errs <<std::endl;
+    return -1;
```
Comment (BillyDonahue) [inline example/readFromStream/readFromStream.cpp]: -1 isn't a good main() exit code. I think it's out of range.
The EXIT_FAILURE would be best here.

Why not print the errs to std::cerr? Why comment it out?

File: example/readFromStream/readFromStream.cpp [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  }
+  // std::cout << jsonRoot <<std::endl;
```
Comment (BillyDonahue) [inline example/readFromStream/readFromStream.cpp]: uncomment this line.
You're suppressing output, but the comments above the example indicate that some output will be produced. There's no need to run silently. Leave these cout/cerr lines in.

File: example/readFromStream/readFromStream.cpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  return 0;
```
Comment (BillyDonahue) [inline example/readFromStream/readFromStream.cpp]: return EXIT_SUCCESS;

File: example/readFromStream/withComment.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+// comment head
+{
+    // comment before
+    "key" : "value"
+    // comment after
+}
+// comment tail
```
Comment (BillyDonahue) [inline example/readFromStream/withComment.json]: weird line terminator.

File: example/readFromString/readFromString.cpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+/*
+parse a string to Value object with CharReaderBuilder class or Reader class
+>g++ readFromString.cpp -ljsoncpp -std=c++11 -o readFromString
+>./readFromString
+colin
+20
+*/
+int main() {
+  std::string strRes = "{\"Age\": 20, \"Name\": \"colin\"}";
+  int nLen = (int)strRes.length();
```
Comment (BillyDonahue) [inline example/readFromString/readFromString.cpp]: auto nLen or just size_t nLen.  No need to cast this to int.

File: example/readFromString/readFromString.cpp [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#include "json/json.h"
+/*
+parse a string to Value object with CharReaderBuilder class or Reader class
+>g++ readFromString.cpp -ljsoncpp -std=c++11 -o readFromString
+>./readFromString
+colin
+20
+*/
+int main() {
+  std::string strRes = "{\"Age\": 20, \"Name\": \"colin\"}";
```
Comment (BillyDonahue) [inline example/readFromString/readFromString.cpp]: raw literals are good for stuff like this.
std::string strRes = R"({"Age": 20, "Name": "colin"})";

File: example/readFromString/readFromString.cpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  const char* pStart = strRes.c_str();
+
+  JSONCPP_STRING errs;
+  Json::Value root;
+
+#if 0 // old way
+    Json::Reader myreader;
+    myreader.parse(strRes,root);
+#else // new way
+  Json::CharReaderBuilder jsonreader;
```
Comment (BillyDonahue) [inline example/readFromString/readFromString.cpp]: misnamed. It's a builder for a reader, not a reader. Also should follow camelcase.

File: example/readFromString/readFromString.cpp [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  Json::CharReader* reader(jsonreader.newCharReader());
```
Comment (BillyDonahue) [inline example/readFromString/readFromString.cpp]: use std::unique_ptr instead of new and delete in the examples.

File: example/readFromString/readFromString.cpp [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+parse a string to Value object with CharReaderBuilder class or Reader class
+>g++ readFromString.cpp -ljsoncpp -std=c++11 -o readFromString
+>./readFromString
+colin
+20
+*/
+int main() {
+  std::string strRes = "{\"Age\": 20, \"Name\": \"colin\"}";
+  int nLen = (int)strRes.length();
+  const char* pStart = strRes.c_str();
```
Comment (BillyDonahue) [inline example/readFromString/readFromString.cpp]: we don't need pStart until the parse() call. Calculate it there or just use the expression directly in the parse call.

File: example/streamWrite/streamWrite.cpp [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+write the Value object to stream
+>g++ streamWrite.cpp -ljsoncpp -std=c++11 -o streamWrite
+>./streamWrite
+{
+    "Age" : 20,
+    "Name" : "robin"
+}
+*/
+int main() {
+  std::string strRes = " ";
```
Comment (BillyDonahue) [inline example/streamWrite/streamWrite.cpp]: Forget these single-space initialized string. This is top-of-function declaration is a C style and C++ doesn't often use it.

Move definition of strRes where it's initialized for real to os.str() below.
But you really don't need it at all.
You can call `std::cout << os.str() << std::endl;`

File: example/streamWrite/streamWrite.cpp [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  Json::Value root;
+  Json::StreamWriterBuilder jsonbuilder;
+  Json::StreamWriter* writer(jsonbuilder.newStreamWriter());
+  std::ostringstream os;
+
+  root["Name"] = "robin";
+  root["Age"] = 20;
+  writer->write(root, &os);
```
Comment (BillyDonahue) [inline example/streamWrite/streamWrite.cpp]: or write directly to &std::cout I guess?

File: example/readFromStream/readFromStream.cpp (hunk 7 above)
Comment (dota17) [reply example/readFromStream/readFromStream.cpp]: I 've tried this , but the EXIT_FAILURE and EXIT_SUCESS are redefined in sysytem's c++ lib.

File: example/readFromString/readFromString.cpp (hunk 14 above)
Comment (dota17) [reply example/readFromString/readFromString.cpp]: Um, I used to use std::unique_ptr, but according to some test cases in main.cpp, I changed it. 
We should also modify them in main.cpp, I guess?

**Comment by dota17** [comment]
Thank you all for all your precious suggests, and now the problem is that the build failed for `undefined reference to XXX`, seems that it can't find the header file, I guess it blames to CMakeLists.txt file, I am not very proficient in its usage. can you give me some suggests on example/CMakeLists.txt file?

File: example/readFromStream/readFromStream.cpp (hunk 7 above)
Comment (BillyDonahue) [reply example/readFromStream/readFromStream.cpp]: What do you mean "redefined"?
They're part of `<cstdlib>`.

Comment (dota17) [reply example/readFromStream/readFromStream.cpp]: It shows that:
```c++
error: "EXIT_SUCCESS" redefined [-Werror]
#define EXIT_SUCCESS 1
In file included from /usr/include/c++/cstdlib:72:0,
                 from /usr/include/c++/5/ext/string_conversions.h:41,
                 from /usr/include/c++/5/bits/basic_string.h:5249,
                         ...
```

Comment (dota17) [reply example/readFromStream/readFromStream.cpp]: OK, I got it, I should delete #define EXIT_SUCESS and EXIT_FAILURE. it's a system marco.

**Comment by dota17** [comment]
@baylesj @BillyDonahue I have modified all change requests, and if there is any thing else to be done, please point it out. 
If this PR is not so necessary, then I can close it because it has been submitted for a long time.

File: example/README.md [hunk 18]
```diff
@@ -0,0 +1,13 @@
+***NOTE*** 
+
+If you get linker errors about undefined references to symbols that involve types in the `std::__cxx11` namespace or the tag
+`[abi:cxx11]` then it probably indicates that you are trying to link together object files that were compiles with different
+values for the _GLIBCXX_USE_CXX11_ABI marco. This commonly happens when linking to a third-party library that was compiled with 
+an older version of GCC. If the third-party library cannot be rebuilt with the new ABI then you need to recompole your code with
+the old ABI,just like :
+**g++ stringWrite.cpp -ljsoncpp -std=c++11 -D_GLIBCXX_USE_CXX11_ABI=0 -o stringWrite**
+
+Not all of uses of the new ABI will cause changes in symbol names,for example a class with a `std::string` member  variable will
```
Comment (baylesj) [inline example/README.md]: Nit: space after comma ("names,for") and weird spacing between "member  variable".

File: example/README.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
+***NOTE*** 
+
+If you get linker errors about undefined references to symbols that involve types in the `std::__cxx11` namespace or the tag
+`[abi:cxx11]` then it probably indicates that you are trying to link together object files that were compiles with different
+values for the _GLIBCXX_USE_CXX11_ABI marco. This commonly happens when linking to a third-party library that was compiled with 
+an older version of GCC. If the third-party library cannot be rebuilt with the new ABI then you need to recompole your code with
+the old ABI,just like :
```
Comment (baylesj) [inline example/README.md]: Nit: space after comma, no space before colon.

File: example/README.md [hunk 20]
```diff
@@ -0,0 +1,13 @@
+***NOTE*** 
+
+If you get linker errors about undefined references to symbols that involve types in the `std::__cxx11` namespace or the tag
+`[abi:cxx11]` then it probably indicates that you are trying to link together object files that were compiles with different
```
Comment (baylesj) [inline example/README.md]: Replace "compiles" with "compiled"

File: example/README.md [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+values for the _GLIBCXX_USE_CXX11_ABI marco. This commonly happens when linking to a third-party library that was compiled with 
+an older version of GCC. If the third-party library cannot be rebuilt with the new ABI then you need to recompole your code with
```
Comment (baylesj) [inline example/README.md]: Add comma: "the new ABI, then you"

File: example/README.md [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+the old ABI,just like :
+**g++ stringWrite.cpp -ljsoncpp -std=c++11 -D_GLIBCXX_USE_CXX11_ABI=0 -o stringWrite**
+
+Not all of uses of the new ABI will cause changes in symbol names,for example a class with a `std::string` member  variable will
+have the same mangled bane whether compiled with the older or new ABI. In order to detect such problems the new types and functions
```
Comment (baylesj) [inline example/README.md]: Replace "bane" with "name"

Comment (baylesj) [inline example/README.md]: Add comma after "In order to detect such problems," since it is a prepositional phrase. See:

https://www.margieholdscourt.com/a-comma-after-a-prepositional-phrase-at-the-beginning-of-a-sentence/

File: example/README.md [hunk 23]
```diff
@@ -0,0 +1,13 @@
...
+asre annotated with the abi_tag attribute,allowing the compiler to warn about potential ABI incompatibilities in code using them.
```
Comment (baylesj) [inline example/README.md]: Replace "asre" with "are"

Add space after "attribute,"

File: example/readFromStream/readFromStream.cpp [hunk 24]
```diff
@@ -0,0 +1,32 @@
+#include "json/json.h"
+#include <fstream>
+#include <iostream>
+/*
+parse from stream,collect comments and capture error info.
```
Comment (baylesj) [inline example/readFromStream/readFromStream.cpp]: Space after stream.

File: example/readFromStream/readFromStream.cpp [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+
```
Comment (baylesj) [inline example/readFromStream/readFromStream.cpp]: Add comment, something like:

"Example usage:"

File: example/readFromStream/readFromStream.cpp [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+    "key" : "value"
+}
+// comment after
+// comment tail
+*/
+int main(int argc, char* argv[]) {
+  Json::Value root;
+  std::ifstream ifs;
+  ifs.open(argv[1]);
+  root.clear();
```
Comment (baylesj) [inline example/readFromStream/readFromStream.cpp]: Isn't the root clear by default? Or no?

File: example/readFromString/readFromString.cpp [hunk 27]
```diff
@@ -0,0 +1,33 @@
+#include "json/json.h"
+#include <iostream>
+/*
+parse a string to Value object with CharReaderBuilder class or Reader class
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: Capitalize "parse"

Comment (baylesj) [inline example/readFromString/readFromString.cpp]: The CharReaderBuilder class cannot parse objects. I would rephrase as:

"Parse a raw string into a Value object using the CharReader class, or the legacy Reader class."

File: example/readFromString/readFromString.cpp [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+$g++ readFromString.cpp -ljsoncpp -std=c++11 -o readFromString
+$./readFromString
+colin
+20
+*/
+int main() {
+  std::string strRes = R"({"Age": 20, "Name": "colin"})";
+  auto nLen = (int)strRes.length();
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: Prefer a modern cast here:

`const int rawJsonLength = static_cast<int>(rawJson);`

File: example/readFromString/readFromString.cpp [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+#include <iostream>
+/*
+parse a string to Value object with CharReaderBuilder class or Reader class
+$g++ readFromString.cpp -ljsoncpp -std=c++11 -o readFromString
+$./readFromString
+colin
+20
+*/
+int main() {
+  std::string strRes = R"({"Age": 20, "Name": "colin"})";
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: This should be const, and a better name, e.g.

`const std::string rawJson = R"({"Age": 20, "Name": "colin"})";`

File: example/readFromString/readFromString.cpp [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+  auto nLen = (int)strRes.length();
+  JSONCPP_STRING err;
+  Json::Value root;
+
+#if 0 // old way
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: Nit: can replace with the following:

```
constexpr bool shouldUseOldWay = true;

if (shouldUseOldWay) {
...
```

File: example/readFromString/readFromString.cpp [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+    Json::Reader myreader;
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: Rename `myreader` to `myReader` or just `reader`

File: example/readFromString/readFromString.cpp [hunk 32]
```diff
@@ -0,0 +1,33 @@
...
+    myreader.parse(strRes,root);
+#else // new way
+  Json::CharReaderBuilder builder;
+  std::unique_ptr<Json::CharReader> const reader(builder.newCharReader());
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: Prefer `const foo` to `foo const`

File: example/readFromString/readFromString.cpp [hunk 33]
```diff
@@ -0,0 +1,33 @@
...
+  if (!reader->parse(strRes.c_str(), strRes.c_str() + nLen, &root, &err)) {
+    std::cout << "error" << std::endl;
+    return EXIT_FAILURE;
+  }
+#endif
+  std::string strName = root["Name"].asString();
```
Comment (baylesj) [inline example/readFromString/readFromString.cpp]: I would prefer not prepending str (e.g. Hungarian notation).

Consider:

`const std::string name = root["Name"].asString();
 const int age = root["Age"].asInt();

File: example/streamWrite/streamWrite.cpp [hunk 34]
```diff
@@ -0,0 +1,22 @@
+#include "json/json.h"
+#include <iostream>
+/*
+write the Value object to stream
```
Comment (baylesj) [inline example/streamWrite/streamWrite.cpp]: Capitalize "write" and change "object to a stream"

File: example/streamWrite/streamWrite.cpp [hunk 35]
```diff
@@ -0,0 +1,22 @@
...
+$./streamWrite
+{
+    "Age" : 20,
+    "Name" : "robin"
+}
+*/
+int main() {
+  Json::Value root;
+  Json::StreamWriterBuilder builder;
+  std::unique_ptr<Json::StreamWriter> const writer(builder.newStreamWriter());
```
Comment (baylesj) [inline example/streamWrite/streamWrite.cpp]: prepend const.

File: example/stringWrite/stringWrite.cpp [hunk 36]
```diff
@@ -0,0 +1,33 @@
...
+{
+    "action" : "run",
+    "data" :
+    {
+        "number" : 1
+    }
+}
+*/
+int main() {
+
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: Replace extra line.

File: example/stringWrite/stringWrite.cpp [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+  Json::Value root;
+  Json::Value t_data;
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: Why `t_data`? Should be `rawData` if anything.

File: example/stringWrite/stringWrite.cpp [hunk 38]
```diff
@@ -0,0 +1,33 @@
...
+
+  root["action"] = "run";
+  t_data["number"] = 1;
+  root["data"] = t_data;
+
+#if 0 // old way
+    Json::FastWriter writer;
+    std::string json_file = writer.write(root);
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: const

File: example/stringWrite/stringWrite.cpp [hunk 39]
```diff
@@ -0,0 +1,33 @@
...
+#else // new way
+  Json::StreamWriterBuilder builder;
+  std::string json_file = Json::writeString(builder, root);
```
Comment (baylesj) [inline example/stringWrite/stringWrite.cpp]: const

File: src/lib_json/CMakeLists.txt [hunk 40]
```diff
@@ -87,7 +87,7 @@ endif()
 # See https://cmake.org/cmake/help/v3.1/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES
 # for complete list of features available
 target_compile_features(jsoncpp_lib PUBLIC
-        cxx_std_11 # Compiler mode is aware of C++ 11.
+#        cxx_std_11 # Compiler mode is aware of C++ 11.
```
Comment (baylesj) [inline src/lib_json/CMakeLists.txt]: Why is this commented out? Don't we want this?

**Comment by baylesj** [comment]
I am glad you did this patch, I definitely want example code for this project. If you can fix up the errors I'm happy to merge it in.

File: example/readFromString/readFromString.cpp (hunk 32 above)
Comment (dota17) [reply example/readFromString/readFromString.cpp]: Do you mean `const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
`?
